engine: plumb a Context into MVCC

tbg opened the pull request:
This will make it much easier to debug those performance issues which can be
triggered through `EXPLAIN (TRACE) <STMT>` in a development setting.

There currently is no good story for different verbosity levels, so the goal is
not to place regular tracepoints around but to have the infrastructure for
ad-hoc debugging performance issues in place. I imagine this could be useful
for issues such as #4386 or #5981.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5255)

<!-- Reviewable:end -->

vivekmenezes commented:
Reviewed 35 of 35 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/replica.go, line 1516 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255#-KCsA1LJzGSlww9fk6Yf:-KCsA1LJzGSlww9fk6Yg:1533838371) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ef3a32fa1363b0ff5307092fd23e892a9b5137a/storage/replica.go#L1516)):*
A `Span` is in scope here. Does it matter that we're passing the context instead of that span?

---

*[storage/replica_command.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255#-KCs6qG_XgUCJOpuZpsC:-KCs6qG_XgUCJOpuZpsD:-1967586653) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ef3a32fa1363b0ff5307092fd23e892a9b5137a/storage/replica_command.go#L52)):*
Didn't you just argue against this in #5184?

I feel like maybe we should be passing the context around and getting the span from the context when we need it. (Didn't I just argue against that in #5184? I think the difference is that the span is a global/contextual thing while the command ID and index are more part of the semantics of the particular function.)

One potential reason to pass the context when all we need is a span: the `log.Infoc` family of methods. We're backing away from structured logging but these methods can still add some extra context to log messages and would let us avoid adding things like store ID and range ID to messages on an ad-hoc basis. I think it would be a good idea to expand their use (and if we're not going to use them more we should just get rid of the few places where we do use them).

---

*[storage/engine/mvcc.go, line 50 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255#-KCs9I4Llf07m8tTtKF_:-KCs9I4Llf07m8tTtKFa:1579667197) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ef3a32fa1363b0ff5307092fd23e892a9b5137a/storage/engine/mvcc.go#L50)):*
This is used for more than just tests and more than just mvcc calls. Should it live in `tracing` instead (and without the last line of the comment)? I'm not quite sure if it's worthwhile to have a special name for `nil`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255#-:-KCsAhNr3ihht6-PA4HM:676834233)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/replica_command.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255#-KCs6qG_XgUCJOpuZpsC:-KCur-vVfnY4nW4euZyV:36412728) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ef3a32fa1363b0ff5307092fd23e892a9b5137a/storage/replica_command.go#L52)):*
I'd prefer to see a `context.Context` passed around everywhere. Seems more idiomatic. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1516 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255#-KCsA1LJzGSlww9fk6Yf:-KCvQ9Pg3UG-lpTBblcA:1364175220) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ef3a32fa1363b0ff5307092fd23e892a9b5137a/storage/replica.go#L1516)):*
That's a good point - it does matter if `SpanFromContext` created the Span. SpanFromContext should also return a new context, but the interface would be awkward - I'll take care of this separately.

---

*[storage/replica_command.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255#-KCs6qG_XgUCJOpuZpsC:-KCvQlf3QMh09rJNuJUE:1337286879) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ef3a32fa1363b0ff5307092fd23e892a9b5137a/storage/replica_command.go#L52)):*
I didn't do that because if we don't have a trace, we want to create one, but we can't do that without a tracer. Sure, we can put that into the context as well, but that doesn't see right. But maybe we should figure out the right way to do it - see comment below.

I've also wanted to hook tracing up with logging, but it didn't seem important enough so far.

---

*[storage/engine/mvcc.go, line 50 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255#-KCs9I4Llf07m8tTtKF_:-KCvQvPCaa7qejs-ac59:-551330682) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ef3a32fa1363b0ff5307092fd23e892a9b5137a/storage/engine/mvcc.go#L50)):*
There's a little bit of a conflict here. On the one hand, we want to avoid nil traces because that means checking in all of the code base that `sp != nil` before calling `sp.LogEvent`.
On the other hand, as this change demonstrates, it's awkward to have a Span everywhere. It's also awkward to try to wrap the Span in something else that fixes these shortcomings.

I think there's a good solution once we've figured out how to combine logging and tracing. I'll suggest a PR shortly - the idea would be that all (usual) logging calls take a context which will log to a span contained in the trace (if any) as well. Additionally, we have `log.Trace(ctx, msg)` (punting on the lazy Printf version for now) which can be used to only log to the trace but not to anywhere else.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on server/node.go:
```diff
@@ -670,6 +670,8 @@ func (n *Node) Batch(ctx context.Context, args *roachpb.BatchRequest) (*roachpb.
 		// back with the response. This is more expensive, but then again,
 		// those are individual requests traced by users, so they can be.
 		if sp.BaggageItem(tracing.Snowball) != "" {
+			sp.LogEvent("delegating to snowball tracing")
```
@nvanbenschoten

tbg committed:
engine: plumb a Context into MVCC

This will make it much easier to debug those performance issues which can be
triggered through `EXPLAIN (TRACE) <STMT>` in a development setting.

There currently is no good story for different verbosity levels, so the goal is
not to place regular tracepoints around but to have the infrastructure for
ad-hoc debugging performance issues in place. I imagine this could be useful
for issues such as #4386 or #5981.

tschottdorf commented:
Resuscitated, PTAL. Don't look at the diff. I suggest skipping lightly over `*_test.go` and to concentrate on `storage/engine`, though I think all changes are trivial (unless I accidentally snuck anything else in).

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

One question is whether the uses of `context.Background()` in non-test code should be `context.TODO()`. See `storage/replica.go`, `storage/stores.go` and `storage/replica_raftstorage.go`.

---

Review status: 0 of 33 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5255#-:-KFOtn5i8988gp-W9jwF:2074715989)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I recall initially wanting to plumb **everywhere**, but it didn't really make a whole lot of sense. There might be the odd Context that should be a `TODO()` but I think I looked at the vast majority and decided they should be `Background()` (i.e. we're not really interested in connecting them to some other upstream context).

tschottdorf commented:
Thanks for the quick review, @petermattis.
